File: Leetcode - Java/144_tree_preorderTraversal.java
// https://leetcode.com/problems/binary-tree-preorder-traversal/

// Definition for a binary tree node.

import java.util.*;

// public class TreeNode {
//     int val;
//     TreeNode left;
//     TreeNode right;
//     TreeNode(int x) { val = x; }
// }
 
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        ArrayList<Integer> preorder = new ArrayList<Integer>();
        if(root != null) {
            preorder.add(root.val);
            preorder.addAll(preorderTraversal(root.left));
            preorder.addAll(preorderTraversal(root.right));
        }
        return preorder; 
    }
}

File: Leetcode-Python-New/144_tree_preorderTraversal.py
# https://leetcode.com/problems/binary-tree-preorder-traversal/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        if root is None:
            return []
        preorderList = []
        self.preorder(root, preorderList)
        return preorderList
    
    def preorder(self, node: TreeNode, preorderList: List[int]) -> None:
        if node is None:
            return
        preorderList.append(node.val)
        self.preorder(node.left, preorderList)
        self.preorder(node.right, preorderList)